File: backend/schemas/module1.py
"""M1 — Intraday Swing Analysis schemas."""
from __future__ import annotations

from typing import Any
from pydantic import BaseModel, Field


class ATRConfig(BaseModel):
    period: int = Field(14, ge=2, le=100)
    multiplier: float = Field(2.0, ge=0.1, le=20.0)
    candle_tf: str = Field("15m", pattern="^(5m|15m|30m)$")
    price_source: str = Field("close", pattern="^(close|hl)$")


class Module1Request(BaseModel):
    instrument: str = Field("NIFTY", pattern="^(NIFTY|BANKNIFTY)$")
    from_date: str          # YYYY-MM-DD
    to_date: str            # YYYY-MM-DD
    pts_to_analyse: int = Field(100, ge=25, le=500)
    option_range: int = Field(3, ge=0, le=10)
    expiry_count: int = Field(2, ge=0, le=5)
    atr: ATRConfig = Field(default_factory=ATRConfig)


class SwingPoint(BaseModel):
    timestamp: str
    price: float
    swing_type: str           # Hill | Valley


class NonQualifyingLeg(BaseModel):
    date: str
    end_date: str
    start_time: str
    end_time: str
    start_price: float
    end_price: float


class LegRow(BaseModel):
    date: str
    end_date: str
    start_time: str
    end_time: str
    leg_type: str             # Hill | Valley
    vix: float | None
    abs_pts: float
    start_price: float
    end_price: float
    # Per-expiry option details: {"exp1": {expiry_date, along_type, against_type, atm_strike, atm_along_name, cheapest_along_strike, cheapest_along_name}, ...}
    expiry_details: dict[str, Any] = Field(default_factory=dict)
    # All generated column values (along/against option prices, metrics, etc.)
    columns: dict[str, Any]


class Module1Result(BaseModel):
    instrument: str
    from_date: str
    to_date: str
    pts_to_analyse: int
    option_range: int
    expiry_count: int
    atr: ATRConfig
    swing_points: list[SwingPoint]
    legs: list[LegRow]
    non_qualifying_legs: list[NonQualifyingLeg]
    # Column metadata: {key: {label, group, expiry_n, strike_label, pct, ...}}
    column_meta: dict[str, Any]
    # OHLC candles for the chart (at ATR candle timeframe)
    chart_candles: list[dict]
    run_id: str               # YYYYMMDD_HHMMSS

File: frontend/src/api/module1.ts
import { apiClient } from './client';

export interface ATRConfig {
  period: number;
  multiplier: number;
  candle_tf: '5m' | '15m' | '30m';
  price_source: 'close' | 'hl';
}

export interface Module1Request {
  instrument: 'NIFTY' | 'BANKNIFTY';
  from_date: string;
  to_date: string;
  pts_to_analyse: number;
  option_range: number;
  expiry_count: number;
  atr: ATRConfig;
}

export interface SwingPoint {
  timestamp: string;
  price: number;
  swing_type: 'Hill' | 'Valley';
}

export interface NonQualifyingLeg {
  date: string;
  end_date: string;
  start_time: string;
  end_time: string;
  start_price: number;
  end_price: number;
}

export interface LegRow {
  date: string;
  end_date: string;
  start_time: string;
  end_time: string;
  leg_type: 'Hill' | 'Valley';
  vix: number | null;
  abs_pts: number;
  start_price: number;
  end_price: number;
  columns: Record<string, number | null>;
}

export interface ColumnMeta {
  label: string;
  group: 'A' | 'B' | 'C' | 'D';
  exp_n?: number;
  strike_label?: string;
  side?: string;
  pct?: number;
}

export interface ChartCandle {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface Module1Result {
  instrument: string;
  from_date: string;
  to_date: string;
  pts_to_analyse: number;
  option_range: number;
  expiry_count: number;
  atr: ATRConfig;
  swing_points: SwingPoint[];
  legs: LegRow[];
  non_qualifying_legs: NonQualifyingLeg[];
  column_meta: Record<string, ColumnMeta>;
  chart_candles: ChartCandle[];
  run_id: string;
}

export async function runModule1(req: Module1Request): Promise<Module1Result> {
  const { data } = await apiClient.post<Module1Result>('/module1/run', req);
  return data;
}

export async function getModule1Result(runId: string): Promise<Module1Result> {
  const { data } = await apiClient.get<Module1Result>(`/module1/results/${runId}`);
  return data;
}
